### ISSUE TITLE
Major performance difference between Windows and Linux

### ISSUE THREAD

**Issue by Tragetaschen** [aspnet/Mvc#2094]
There is a major performance difference using the same binaries between Windows and Mono. I have bundled my beta3-based application and run the application with the `*.cmd` and `*.sh` respectively.

On Windows, the first request for Home/Index takes approximately 1000ms, every subsequent one is in the 10-20ms range. On Linux, the first request takes about 25000ms (really, that's 25s) and every subsequent request still 3500ms. According to F12 tools the time is spent waiting for the response.

I can see this difference with both `Kestrel` and `Nowin.vNext`, so I would rule out the HTTP server component. Also, I don't see any notable difference in serving static content files which both systems handle in the two-digit-millisecond range.

The Windows system is a Core i7 Surface Pro 3.
The Linux is an embedded ARMv7 with 792 MHz.

I would expect a difference, but it's several orders of magnitude too large to result from the slower system speed.

Any advice where and how to start looking?

**Comment by Tragetaschen** [comment]
A simple content action (`return "Hello";`) show 4 ms / 28 ms on Windows / Linux.

If I remove `_ViewStart.cshtml`, I get 600ms / 15,000ms for the first page view and 7ms / 43ms for subsequent requests.

**Comment by Tragetaschen** [comment]
Without precompilation, without `_ViewStart.cshtml` I get 6000/15000 for the first request and fast as above for subsequent requests.

So I can say:
- The precompilation done on Windows during `bundle` is not picked up on Linux.
- An existing `_ViewStart.cshtml` file seems to invalidate the Razor compilation result on Linux for every request.

**Comment by Tragetaschen** [comment]
The numbers are the same for a blank `ASP.NET 5 Preview Starter Web` with added `Kestrel`

**Comment by Tragetaschen** [comment]
I've found two issues and from what I can see, both are already fixed on `dev` (albeit accidentally :-) ):

[First](https://github.com/Tragetaschen/Mvc/commit/ca32626c37e82a361d94eadee517d42bc0d3b8a6), `_ViewStart.cshtml` files in beta3 are associated with a page's cache entry using the normalized path instead of the native one, so enumerating the possible locations fails and the cache property is `null` everywhere resulting in compilation for each page request. There might be a similar issue [on current dev](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs#L201), but I haven't tried any of this.

[Second](https://github.com/Tragetaschen/Mvc/commit/92005a13964878da7fec492e49c947a3867191ea), during pre-compilation in beta3, the each file's hash is not stored in the generated cache entries and doesn't land in the generated code. During the introduction of `IMemoryCache`, [this has been fixed](https://github.com/aspnet/Mvc/commit/59db8143dae4eabff79d56dcd387839cfd0d3436#diff-06929e243aba542b5cfbe01f30aae7a8R178). The reason I could see this was because the Windows machine was running on CET and the Linux machine on UTC, so the timestamps didn't match when evaluating the cache entries.
